### PR TITLE
Skybox

### PR DESCRIPTION
**Pull request by Dhruv-0-Arora**
### Description
Created a spherical skybox using glsl shaders. The skybox is currently implemented in the SceneRenderer.ts file and uses the glsl shaders located in @/shaders. In /public/ there is the custom starry_sky.png used to create the stars. These will most likely be removed or changed.

### Key Features:
- Starry Skybox: Uses custom starry image as the sky for the spherical skybox.
- Color Mixing: Shaders are utilized to blend colors, adding depth and realism. The shaders use the lygia package to incorporate a Perlin noise algorithm which allows for the generation of different patterns.
- Theme Integration Ready: The implementation has the ability to utilize the current theme. However, it currently uses hard-coded colors to create the skybox.

> [!WARNING]
> The renderer for the robot has been commented out in this code. There are also numerous changes to the tsconfig & vite config.

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1668)

### CODE REVIEW

File: fission/public/skycube_1/skyrender0001.bmp
Comment (HunterBarclay) [inline fission/public/skycube_1/skyrender0001.bmp:1]: Please remove all the texture files. If we do end up using a texture in the future, we'll include it in the assetpack.

File: fission/src/shaders/fragment.glsl
Comment (HunterBarclay) [inline fission/src/shaders/fragment.glsl:1]: Adjust the formatting on this file to be consistent and bundle code you think is worth saving in large comments at the end of the file.

Comment (HunterBarclay) [reply fission/src/shaders/fragment.glsl:1]: And detail the code you save

File: fission/src/systems/scene/SceneRenderer.ts [hunk 1]
```diff
@@ -80,6 +85,38 @@ class SceneRenderer extends WorldSystem {
         ground.receiveShadow = true;
         ground.castShadow = true;
         this._scene.add(ground);
+
+        // skybox
+        const currentTheme = (window as any).getTheme();
+        console.log('Current Theme:', currentTheme['Background']['color']['r']);
+        
+        const textureLoader = new THREE.TextureLoader();
+        const cloudTexture = textureLoader.load('./starry_sky.png');
```
Comment (HunterBarclay) [inline fission/src/systems/scene/SceneRenderer.ts]: Remove cloud texture

File: fission/package.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
   "type": "module",
   "scripts": {
     "dev": "vite",
-    "build": "tsc && vite build",
+    "build": "tsc && vite build --base=/",
```
Comment (HunterBarclay) [inline fission/package.json]:
```suggestion
    "build": "tsc && vite build",
```

**Review by HunterBarclay — CHANGES_REQUESTED** [review]
Could you remove the stars?

**Comment by Dhruv-0-Arora** [comment]
Resolved all changes

File: fission/src/shaders/fragment.glsl [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	// vec2 uv = absPos.x == maxCoord ? vec2(vPosition.z, vPosition.y) / maxCoord :
+	// 		absPos.y == maxCoord ? vec2(vPosition.x, vPosition.z) / maxCoord :
+	// 								vec2(vPosition.x, vPosition.y) / maxCoord;
+	// adjustedUV = (uv + 1.0) / 2.0;
+	// adjustedUV.x = vPosition.x < 0.0 || vPosition.z < 0.0 ? 1.0 - adjustedUV.x : adjustedUV.x;
+	// adjustedUV.y = vPosition.y < 0.0 ? 1.0 - adjustedUV.y : adjustedUV.y;
+
+	// vec4 imageColor = texture2D(uTexture, adjustedUV);
+	// vec3 colorMix = mix(vec3(0.0, 0.0, 0.0), blueAccent, noiseValue*0.7);
+	// vec3 skyColor = mix(colorMix, imageColor.rgb, imageColor.a);
```
Comment (BrandonPacewic) [inline fission/src/shaders/fragment.glsl]: I would prefer to leave large chunks of commented out code in a separate branch if the goal is to reference back to it later but that would be up to @HunterBarclay.

File: fission/src/systems/scene/SceneRenderer.ts [hunk 4]
```diff
@@ -141,6 +172,52 @@ class SceneRenderer extends WorldSystem {
...
+    //     };
+    //     image.src = url;
+    //     console.log('Loaded Texture:', image.src);
+
+    //     return texture!;
+    // }
+
+    // public isPowerOf2(value: number): boolean {
+    //     return (value & (value - 1)) === 0;
+    // }
```
Comment (BrandonPacewic) [inline fission/src/systems/scene/SceneRenderer.ts]: Same as previous comment.

I would prefer to leave large chunks of commented out code in a separate branch if the goal is to reference back to it later but that would be up to @HunterBarclay.

File: fission/src/shaders/fragment.glsl [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    vec3 blendedColor = mix(horizonColor, skyColor, tHorizon);
+    vec3 finalColor = mix(voidColor, blendedColor, tVoid);
+
+    return finalColor;
+}
+
+// some calculations
+float func(float x) {
+	return x*x*(3.0-2.0*x);
+}
```
Comment (BrandonPacewic) [inline fission/src/shaders/fragment.glsl]: What kind of calculations? 🤔

File: fission/src/shaders/fragment.glsl [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+uniform float rColor;
+uniform float gColor;
+uniform float bColor;
+
+varying vec3 vPosition;
+varying vec2 vUv;
+
+#define PI 3.14159265358979323846264338327
+
+#include "/node_modules/lygia/generative/snoise.glsl" // this is wack
```
Comment (BrandonPacewic) [inline fission/src/shaders/fragment.glsl]: I agree, this definitely looks wack. Is this the best way you could find to go about doing this?

File: fission/src/Synthesis.tsx [hunk 7]
```diff
@@ -166,25 +166,25 @@ function Synthesis() {
...
+            //         console.error(`Assembly Parser produced significant errors for '${miraAssembly.info!.name!}'`);
+            //         return;
+            //     }
                 
-                const mirabufSceneObject = new MirabufSceneObject(new MirabufInstance(parser));
-                World.SceneRenderer.RegisterSceneObject(mirabufSceneObject);
-            })();
+            //     const mirabufSceneObject = new MirabufSceneObject(new MirabufInstance(parser));
+            //     World.SceneRenderer.RegisterSceneObject(mirabufSceneObject);
+            // })();
```
Comment (BrandonPacewic) [inline fission/src/Synthesis.tsx]: I see you mentioned that this code was commented out in the PR description, was there a reason for this?

**Review by BrandonPacewic — COMMENTED** [review]
Some minor clarifying points. When you get the chance try to resolve the merge conflicts (unless @HunterBarclay does it for you 🙃). If you need any help with that feel free to grab me!

File: fission/src/Synthesis.tsx (hunk 7 above)
Comment (Dhruv-0-Arora) [reply fission/src/Synthesis.tsx]: Hunter had commented it so it would run faster on my computer

Comment (BrandonPacewic) [reply fission/src/Synthesis.tsx]: 👍

File: fission/src/shaders/fragment.glsl
Comment (BrandonPacewic) [reply fission/src/shaders/fragment.glsl:1]: @Dhruv-0-Arora Moved the commented out code sections into his own fork on another branch.

File: fission/src/systems/scene/SceneRenderer.ts [hunk 8]
```diff
@@ -68,11 +73,34 @@ class SceneRenderer extends WorldSystem {
...
-        ground.castShadow = true
-        this._scene.add(ground)
+        const ground = new THREE.Mesh(new THREE.BoxGeometry(10, 1, 10), this.CreateToonMaterial(GROUND_COLOR));
+        ground.position.set(0.0, -2.0, 0.0);
+        ground.receiveShadow = true;
+        ground.castShadow = true;
+        this._scene.add(ground);
+
+        // skybox
+        const currentTheme = (window as any).getTheme();
```
Comment (BrandonPacewic) [inline fission/src/systems/scene/SceneRenderer.ts]: ESLint wants a different type here instead of `any`. Any chance we can change this?

Comment (HunterBarclay) [reply fission/src/systems/scene/SceneRenderer.ts]: @Dhruv-0-Arora What is this `getTheme()` function?

Comment (HunterBarclay) [reply fission/src/systems/scene/SceneRenderer.ts]: If you're trying to get the background to update to the scene, I would build out a component that, when re-rendered, updates the uniforms of the skybox

Comment (Dhruv-0-Arora) [reply fission/src/systems/scene/SceneRenderer.ts]: the get theme function is supposed to get the currentTheme. Right now its just black

Comment (Dhruv-0-Arora) [reply fission/src/systems/scene/SceneRenderer.ts]: I think I probably need to fix it

**Comment by Dhruv-0-Arora** [comment]
@HunterBarclay @BrandonPacewic  do the changes to skybox fix everything for you?

**Review by HunterBarclay — APPROVED** [review]
Looking good for now. We'll be changing up theme and what not pretty significantly so we'll address changes then

File: fission/src/systems/scene/SceneRenderer.ts [hunk 9]
```diff
@@ -137,6 +161,21 @@ class SceneRenderer extends WorldSystem {
             gradientMap: gradientMap,
         })
     }
+
+    /* 
+     * Updates the skybox colors based on the current theme
+
+     * @param currentTheme: current theme from ThemeContext.useTheme()
```
Comment (BrandonPacewic) [inline fission/src/systems/scene/SceneRenderer.ts:168]: Probably best to have this than not. I would prefer that we standardize whether or not we want these kinds of doc strings at some point but that's a moot point anyways.

Looks good to me!

**Review by BrandonPacewic — APPROVED** [review]

**Pull request merged**